### ISSUE TITLE
Switch between multiple sets of bindings

### ISSUE THREAD

**Issue by jacobtomlinson** [Textualize/textual#1302]
It would be great to support binding modes in my textual apps. Similar to command mode and edit mode in vim I'd like to be able to switch out different sets of bindings.

I see I can have different bindings associated with different screens but I'd like to be able to change between binding sets within one screen.

I tried updating the `BINDINGS` constant within an action with a new set of bindings but it didn't seem to have any effect (and modifying a constant feels icky).

**Comment by davep** [comment]
I what sort of situations are you finding you want to swap bindings within a screen? What sort of context causes the wish to swap around? I ask only because, last week, I sort of found myself wanting to do the same but I realise that what I really wanted to be doing was [to have relevant bindings at the widget level](https://textual.textualize.io/blog/2022/11/26/on-dog-food-the-original-metaverse-and-not-being-bored/#dont-put-all-your-bindings-in-one-place).

**Comment by jacobtomlinson** [comment]
Oh interesting I didn't realise you could add bindings to widgets! So would you have different bindings when different widgets are in focus?

I have about 8 widgets on my screen, but I only want two of them to be focusable, and one or other should always bein focus.

**Comment by davep** [comment]
> So would you have different bindings when different widgets are in focus?

Yup, that's what I did with [this app](https://github.com/davep/unbored). Mostly I don't want any bindings other than a couple of very general ones, but when focus is within one of the activity displays, [I have a whole bunch of bindings on that widget](https://github.com/davep/unbored/blob/452afa472ece34c69da7176f7b0b6e87076527e0/unbored/activity.py#L45). Note too that the bindings are in effect if any of that widget's children are focused.

More generally, this is how bindings work in Textual. There's a slight aside I won't get into here, but in effect everything is a widget and a binding declared for a widget is good for that widget and all child widgets. Define for your app? That's good for all screens and widgets within. Defined for one screen? That's good for all widgets within. Defined for a particular widget? That's good for that widget and its children.

**Comment by jacobtomlinson** [comment]
Ok that makes sense. In that case I have two widgets that I want different (overlapping) bindings on. If neither of them is in focus I want one of them to be kept in focus. I.e none of the other widgets should be focusable. Do you have thoughts on a "correct" way to do that?

**Comment by davep** [comment]
What are the other widgets such that they can be focused but you don't want them to have focus? Normally I'd only use a focusable widget if I want it to have focus, otherwise I'd use something different that can't have focus.

**Comment by jacobtomlinson** [comment]
Heh I guess this is getting to the core of my misunderstanding. I assumed all widgets were always focusable (but I couldn't seem to get focus working either). From your question I'm guessing widget focus is opt in somehow which makes sense why I couldn't get my widget to focus. I can't figure out from the docs how to make a "focusable widget".

Maybe to rephrase my question. How do I make two of my widgets focusable? And how do I ensure that I can't be in a fully blurred state (one widget must always be in focus)?

_(Feel free to redirect me to the forum if that's a better place for this question)_

**Comment by davep** [comment]
> Feel free to redirect me to the forum if that's a better place for this question

Aye, I think this is turning more into a Q&A, I'll move this to there.
